libtransmission/session.h

```diff
@@ -1100,7 +1114,10 @@ struct tr_session
     // depends-on: open_files_
     tr_torrents torrents_;
 
-    // depends-on: settings_, session_thread_, torrents_
+    // depends-on: settings_, session_thread_, timer_maker_, web_
+    tr_global_ip_cache global_ip_cache_{ *this };
```

Comment:
I don't love the idea of `tr_global_ip_cache` knowing about `tr_session`. It's been a lot of work to try & decouple some of these components, and once a class knows about `tr_session` basically no decoupling is possible. :disappointed: 

looks like the cache uses session for:

- its TimerMaker. This could accomplished passing a TimerMaker& into the constructor, instead
- `session_.fetch()`, in `update_global_addr()`. This could be accomplished by passing a tr_web& into that method instead.
- session_.settings_.bind_address_ipv4 and session_.settings_.bind_address_ipv6. This could be accomplished by passing them into the constructor and saving them as private member variables.
- having `TR_ASSERT(am_in_session_thread)` -- which, TBH, I'm not sure the cache code needs to know about the session thread, so maybe these can be removed?

Reply:
I will take my sweet time on this. When I think it through, it seems to me that `tr_global_ip_cache` will have to initialise in at least 3 separate stages after this change.

Fully expect to bump into obstacles on this one.

Reply:
You know this PR better than I do, but I'm not seeing the three stages -- could you walk me through them?

Reply:
Oh a second thought, do you know if anything will use `tr_session::publicAddress()` before `tr_session::setSettings()` is called in `tr_session::initImpl()`?

If there are none (and it makes sense that there are none), then I just need to construct `global_ip_cache_` inside `tr_session::initImpl()` with `tr_web&` and `TimerMaker&` as parameter, and then pass `settings_.bind_address_ipv*` inside `tr_session::setSettings()`.